## Code Review: PR #10074

### Summary
Solid infra PR adding a one-click rollback path for the Rust runner fleet. Two commits: the original implementation (`d3b4c16`) and a clean P2-hardening follow-up (`7fc9363`) that resolved both Phase 4 timeout and Phase 7 `bin_dir` precheck concerns. No TypeScript, no tests, no DB code — entirely YAML (GitHub Actions + Ansible) plus embedded bash.

### Key Findings

#### Critical Issues (P0)
None.

#### High Priority (P1)
1. **Hardcoded `api_url=https://www.vm0.ai`** in `.github/workflows/rollback-runner.yml:164`. The forward path bakes `api_url` into `runner.yaml` during build, so promote doesn't pass it explicitly — but the rollback re-runs `runner config` on the install path. Hardcoding in the workflow diverges from the rest of the infra where host-level vars/secrets are source of truth. Prefer a repo/env var (e.g. `${{ vars.VM0_API_URL }}`) so staging/prod can diverge and future domain changes don't require a workflow edit. (Bad-smell #6/#11.)
2. **Missing cross-workflow mutex** with `release-please.yml` (uses `${{ github.workflow }}-${{ github.ref }}` while this workflow uses `runner-deploy`). Already acknowledged inline as a TODO tied to #9746 constraint #6 — does not block merge but worth tracking, because the race creates the exact split-version state this workflow is meant to cure.

#### Medium Priority (P2)
- **RESOLVED in 7fc9363**: Phase 7 now asserts `{{ bin_dir }}/runner` exists + is executable unconditionally — handles the short-circuit path (target already Running at Phase 1) where Phase 5/6 didn't pre-stat the binary. `.executable` check exceeds the original suggestion.
- **RESOLVED in 7fc9363**: Phase 4 "Wait for target to go inactive" bumped 30 retries → 60 (60s → 120s), with inline comment tying the bound to systemd's `TimeoutStopSec=5min`.
- Minor stylistic: Slack notification steps duplicate payload across start/success/failure (~60 lines YAML). Mirrors `release-please.yml`, so this is intentionally consistent — YAGNI says leave it, flagging only for awareness.

### Testing Review

#### Coverage
PR adds 2 new files + modifies 1, all infrastructure (GitHub Actions YAML + Ansible playbook):
- `.github/workflows/rollback-runner.yml` (209 lines, new) — no unit tests; workflows are validated by execution
- `ansible/playbooks/rollback-runner.yml` (454 lines, new) — no playbook tests; manual staging validation per repo convention
- `.github/workflows/release-please.yml` (+61/-18) — dashboard job refactor, covered by existing release-please runs

#### Convention Compliance
No test files in diff; no convention violations to flag. Repo has no precedent for automated testing of Ansible playbooks or workflow YAML.

#### Manual Test Plan
PR body includes an 8-item staging checklist covering: dashboard render, semver validation, `confirmed_compatible` gate, happy path (short-circuit), drain path (SIGUSR2 resume), install path (gc-evicted binary), emergency mode (FC VM cleanup), and Slack notifications. Thorough and well-scoped.

#### Testing Verdict: Not Applicable
Changes are infra/CI only. No test requirement per project conventions. Manual staging test plan is documented.

### Bad Smell Analysis
- Mock violations: N/A (no TS)
- Defensive programming: 0 violations (shell `exit 1` on bad input is fail-fast)
- Hardcoded URLs (#6/#11): **1** — `api_url=https://www.vm0.ai`
- Fallback patterns: 0 — the one `failed_when: false` in Phase 3 is necessary because the Draining→Stopping race is expected and handled explicitly downstream
- Lint suppressions: 1 benign `shellcheck disable=SC2029` (pre-existing, standard SSH pattern)
- YAGNI adherence: Good — `rollback_mode` is a 2-value enum, not a policy object; no speculative abstractions
- Over-engineering: None — 8-phase state machine is justified by heterogeneous fleet states converging to one target

### Recommendations
- [ ] **P1**: Replace hardcoded `api_url=https://www.vm0.ai` with `${{ vars.VM0_API_URL }}` in `.github/workflows/rollback-runner.yml:164`
- [ ] **P1 (tracked)**: Cross-workflow mutex with release-please deploy — already noted inline
- [ ] Execute all 8 staging test-plan checkboxes before enabling in prod

### Positive Highlights
- Excellent inline documentation — SIGUSR2 race, concurrency TODO, URL-param-prefill limitation all explained
- State-machine model well-drawn; Phase 3 confluence-point comment especially clear
- Strict semver regex on `target_version` (injection-safe); hard `confirmed_compatible` operator gate
- Preflight asset check fails fast before touching hosts
- Phase 5 "always re-run setup/build/config" matches `build-runner.yml` exactly — same contract both directions
- `--protect-version` passed to gc so rollback target can't be evicted by its own cleanup
- Both drain and stop routes exit via the runner CLI → FC VMs torn down cleanly
- `--force` in emergency mode correctly documented as "SIGTERM via systemctl, runner still runs its shutdown path" — not SIGKILL

### Verdict
**LGTM with one P1 nit** — address the hardcoded `api_url` either in this PR or a follow-up. The cross-workflow mutex is already tracked via inline TODO. No blockers.

---
*Full review details: `codereviews/20260419-pr10074/`*
*Testing standards: `docs/testing.md`*

--- a/.github/workflows/release-please.yml
+++ b/.github/workflows/release-please.yml
@@ -1243,6 +1243,7 @@ jobs:
           AWS_METAL_RUNNER_USER: ${{ vars.AWS_METAL_RUNNER_USER }}
           OFFICIAL_RUNNER_SECRET: ${{ secrets.OFFICIAL_RUNNER_SECRET }}
           RUNNER_VERSION: ${{ needs.release-please.outputs.runner_rs_version }}
+          API_URL: ${{ vars.VM0_API_URL }}
           GRAFANA_CLOUD_PROMETHEUS_URL: ${{ secrets.GRAFANA_CLOUD_PROMETHEUS_URL }}
           GRAFANA_CLOUD_PROMETHEUS_USER: ${{ secrets.GRAFANA_CLOUD_PROMETHEUS_USER }}
           GRAFANA_CLOUD_API_KEY: ${{ secrets.GRAFANA_CLOUD_API_KEY }}
@@ -1275,7 +1276,7 @@ jobs:
             playbooks/build-runner.yml \
             -e "ansible_user=${AWS_METAL_RUNNER_USER}" \
             -e "official_runner_secret=${OFFICIAL_RUNNER_SECRET}" \
-            -e "api_url=https://www.vm0.ai" \
+            -e "api_url=${API_URL}" \
             -e "runner_version=${RUNNER_VERSION}" \
             -e "r2_account_id=${R2_ACCOUNT_ID:-}" \
             -e "r2_access_key_id=${R2_ACCESS_KEY_ID:-}" \

--- a/.github/workflows/rollback-runner.yml
+++ b/.github/workflows/rollback-runner.yml
@@ -151,6 +151,7 @@ jobs:
           GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
           RUNNER_VERSION: ${{ inputs.target_version }}
           ROLLBACK_MODE: ${{ inputs.rollback_mode }}
+          API_URL: ${{ vars.VM0_API_URL }}
         run: |
           cd ansible
           ansible-playbook \
@@ -161,7 +162,7 @@ jobs:
             -e "rollback_mode=${ROLLBACK_MODE}" \
             -e "sentry_dsn=${SENTRY_DSN_RUNNER}" \
             -e "official_runner_secret=${OFFICIAL_RUNNER_SECRET}" \
-            -e "api_url=https://www.vm0.ai" \
+            -e "api_url=${API_URL}" \
             -e "github_token=${GH_TOKEN}" \
             -v
 

--- a/ansible/playbooks/rollback-runner.yml
+++ b/ansible/playbooks/rollback-runner.yml
@@ -82,14 +82,22 @@
     # Typos like rollback_mode=greaceful silently no-op both Phase 7 drain
     # branches, leaving the fleet in a mixed-version state while the play
     # reports success. Catch that upfront.
+    # Semver regex is enforced at the workflow layer too (rollback-runner.yml
+    # Validate target_version format), but we re-check here as defense in
+    # depth for direct `ansible-playbook` invocations: `runner_version`
+    # flows into systemd unit names and `command:` tasks (Ansible splits
+    # on whitespace), so a malformed value containing spaces or shell
+    # metachars could inject extra CLI flags.
     - name: Validate required inputs
       assert:
         that:
           - runner_version is defined and runner_version | length > 0
+          - runner_version is match('^[0-9]+\.[0-9]+\.[0-9]+$')
           - rollback_mode is defined and rollback_mode in ["graceful", "emergency"]
         fail_msg: >-
-          Missing or invalid inputs. Required: `runner_version` (e.g. 0.81.0)
-          and `rollback_mode` (graceful | emergency). Got
+          Missing or invalid inputs. Required: `runner_version` in semver
+          form (e.g. 0.81.0, no leading v) and `rollback_mode` (graceful |
+          emergency). Got
           runner_version='{{ runner_version | default('<undefined>') }}',
           rollback_mode='{{ rollback_mode | default('<undefined>') }}'.
 
@@ -427,15 +435,24 @@
       register: running_runners
       changed_when: false
 
+    # `failed_when: false` tolerates non-zero exits from individual
+    # runners — e.g. a runner already in `draining` (operator pre-drained
+    # it) rejects a second `service drain` with non-zero; a runner mid-
+    # teardown rejects `service stop`. We don't want one stuck unit to
+    # abort the fleet-wide rollback; Phase 8's final `runner doctor` on
+    # the target is the real convergence check. Per-item errors are
+    # still visible in Ansible's per-host log.
     - name: Drain non-target runners (graceful)
       command: "{{ bin_dir }}/runner service drain --name {{ item }}"
       loop: "{{ running_runners.stdout_lines | default([]) }}"
       when: item != runner_name and rollback_mode == "graceful"
+      failed_when: false
 
     - name: Stop non-target runners (emergency)
       command: "{{ bin_dir }}/runner service stop --name {{ item }} --force"
       loop: "{{ running_runners.stdout_lines | default([]) }}"
       when: item != runner_name and rollback_mode == "emergency"
+      failed_when: false
 
     # ========================================
     # Phase 8: Garbage collect + final health check

